hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientReliableTopicProxy.java

```diff
@@ -197,6 +213,125 @@ public LocalTopicStats getLocalTopicStats() {
+                        }
+                    }
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
+            }
+        } catch (RuntimeException e) {
+            throw e;
+        } catch (Exception e) {
+            throw (RuntimeException) peel(e, null,
+                    "Failed to publish messages: " + payload + " to topic:" + getName());
+        }
+    }
+
+    @Override
+    public CompletionStage<Void> publishAllAsync(@Nonnull Collection<? extends E> payload) {
+        checkNotNull(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
+        checkNoNullInside(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
+        InternalCompletableFuture<Void> returnFuture = new InternalCompletableFuture<>();
+
+        List<ReliableTopicMessage> messages = payload.stream()
+                .map(m -> new ReliableTopicMessage(toData(m), null))
+                .collect(Collectors.toList());
+        switch (overloadPolicy) {
+            case ERROR:
+                ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL)
+                        .whenCompleteAsync((id, t) -> {
+                            if (t != null) {
+                                returnFuture.completeExceptionally(t);
+                            }
+                            if (id == -1) {
+                                returnFuture.completeExceptionally(new TopicOverloadException(
+                                        "Failed to publish messages: " + payload + " on topic:" + getName()));
+                            }
+                            returnFuture.complete(null);
+                        });
+
+                break;
+            case DISCARD_OLDEST:
+                ringbuffer.addAllAsync(messages, OverflowPolicy.OVERWRITE)
+                        .whenCompleteAsync((id, t) -> {
+                            if (t != null) {
+                                returnFuture.completeExceptionally(t);
+                            }
+                            returnFuture.complete(null);
+                        });
+                break;
+            case DISCARD_NEWEST:
+                ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL)
+                        .whenCompleteAsync((id, t) -> {
+                            if (t != null) {
+                                returnFuture.completeExceptionally(t);
+                            }
+                            returnFuture.complete(null);
+                        });
+                break;
+            case BLOCK:
+                addAsyncAndBlock(payload, returnFuture, messages, INITIAL_BACKOFF_MS);
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
+        }
+
+        return returnFuture;
+    }
+
+    private void addAsyncAndBlock(@Nonnull Collection<? extends E> payload,
```

Comment:
Do you have a test where this is exercised? I'd like to see which threads are involved in this execution and if it works nicely.

Reply:
Hello @mmedenjak, I've finished implementing some extended tests. Please see `com.hazelcast.client.topic.ClientReliableTopicOverloadTest` (and in special the extended abstract class `com.hazelcast.topic.impl.reliable.TopicOverloadAbstractTest`)